ポートフォワーディング切断中に例外が発生する

- エラーメッセージ: `Unhandled exception`
- 例外: `System.InvalidOperationException: コレクションが変更されました。列挙操作は実行されない可能性があります。`
- 原因: 切断中に再度切断処理が実行されることでフォワーディングポートを管理しているコレクションが変更されている。

### エラーログ

```
2020-05-04 16:39:32.5257 [ERROR] Unhandled exception. - kenzauros.RHarbor.App.TaskScheduler_UnobservedTaskException() System.InvalidOperationException: コレクションが変更されました。列挙操作は実行されない可能性があります。
   場所 System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
   場所 System.Collections.Generic.List`1.Enumerator.MoveNextRare()
   場所 System.Linq.Enumerable.WhereSelectEnumerableIterator`2.MoveNext()
   場所 kenzauros.RHarbor.ViewModels.SSHConnectionViewModel.<Disconnect>d__16.MoveNext() 場所 C:\Repos\private\rharbor\RHarbor\ViewModels\ConnectionViewModels\SSHConnectionViewModel.cs:行 327
--- 直前に例外がスローされた場所からのスタック トレースの終わり ---
   場所 System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   場所 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   場所 System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   場所 kenzauros.RHarbor.ViewModels.ConnectionViewModel`1.<DisconnectRequiredConnection>d__14.MoveNext() 場所 C:\Repos\private\rharbor\RHarbor\ViewModels\ConnectionViewModels\ConnectionViewModel(T).cs:行 138
--- 直前に例外がスローされた場所からのスタック トレースの終わり ---
   場所 System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   場所 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   場所 System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   場所 kenzauros.RHarbor.ViewModels.RDPConnectionViewModel.<Disconnect>d__7.MoveNext() 場所 C:\Repos\private\rharbor\RHarbor\ViewModels\ConnectionViewModels\RDPConnectionViewModel.cs:行 71
--- 直前に例外がスローされた場所からのスタック トレースの終わり ---
   場所 System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   場所 System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   場所 System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   場所 kenzauros.RHarbor.ViewModels.RDPConnectionViewModel.<<Connect>b__6_1>d.MoveNext() 場所 C:\Repos\private\rharbor\RHarbor\ViewModels\ConnectionViewModels\RDPConnectionViewModel.cs:行 45
```

--- a/RHarbor/ViewModels/ConnectionViewModels/ConnectionViewModel(T).cs
+++ b/RHarbor/ViewModels/ConnectionViewModels/ConnectionViewModel(T).cs
@@ -46,7 +46,7 @@ namespace kenzauros.RHarbor.ViewModels
 
         public ConnectionViewModel()
         {
-            DisconnectCommand = IsConnected.CombineLatest(IsConnecting, (a, b) => a || b).ToReactiveCommand();
+            DisconnectCommand = IsConnected.CombineLatest(IsConnecting, (a, b) => a || b).ToAsyncReactiveCommand();
             DisconnectCommand.Subscribe(() => Disconnect()).AddTo(Disposable);
 
             DisplayText = ConnectionName.CombineLatest(Host, Port, IsConnecting,

--- a/RHarbor/ViewModels/ConnectionViewModels/ConnectionViewModel.cs
+++ b/RHarbor/ViewModels/ConnectionViewModels/ConnectionViewModel.cs
@@ -18,7 +18,7 @@ namespace kenzauros.RHarbor.ViewModels
         public IConnectionViewModel Parent { get; set; }
         public ObservableCollection<IConnectionViewModel> Children { get; set; } = new ObservableCollection<IConnectionViewModel>();
 
-        public ReactiveCommand DisconnectCommand { get; set; }
+        public AsyncReactiveCommand DisconnectCommand { get; set; }
 
         /// <summary>
         /// Whether this connection is root or not.

--- a/RHarbor/ViewModels/ConnectionViewModels/IConnectionViewModel.cs
+++ b/RHarbor/ViewModels/ConnectionViewModels/IConnectionViewModel.cs
@@ -16,7 +16,7 @@ namespace kenzauros.RHarbor.ViewModels
         IConnectionViewModel Parent { get; set; }
         ObservableCollection<IConnectionViewModel> Children { get; set; }
 
-        ReactiveCommand DisconnectCommand { get; set; }
+        AsyncReactiveCommand DisconnectCommand { get; set; }
 
         /// <summary>
         /// Whether this connection is root or not.

--- a/RHarbor/ViewModels/ConnectionViewModels/SSHConnectionViewModel.cs
+++ b/RHarbor/ViewModels/ConnectionViewModels/SSHConnectionViewModel.cs
@@ -191,6 +191,7 @@ namespace kenzauros.RHarbor.ViewModels
             {
                 KeepAliveInterval = TimeSpan.FromMilliseconds(ConnectionInfo.KeepAliveInterval)
             };
+            client.ConnectionInfo.Timeout = TimeSpan.FromSeconds(2);
             var ev = new ManualResetEventSlim(false);
             void hostKeyEventHandler(object o, HostKeyEventArgs e)
             {

--- a/RHarbor/ViewModels/ConnectionViewModels/SSHConnectionViewModel.cs
+++ b/RHarbor/ViewModels/ConnectionViewModels/SSHConnectionViewModel.cs
@@ -319,31 +319,33 @@ namespace kenzauros.RHarbor.ViewModels
         /// <returns></returns>
         public override async Task Disconnect()
         {
+            if (SshClient == null) return;
+            this.WriteLog("Disconnecting...");
+            var ports = ForwardedPorts?.ToList();
+            if (ports?.Count > 0)
+            {
+                foreach (var pf in ports)
+                {
+                    try
+                    {
+                        await pf.Disconnect();
+                    }
+                    catch (Exception ex)
+                    {
+                        MyLogger.Log("Failed to close a forwarded port.", ex);
+                    }
+                }
+                ports.ForEach(x => Children.Remove(x));
+                this.WriteLog("Closed All Forwarded Ports.");
+            }
             if (SshClient != null)
             {
-                this.WriteLog("Disconnecting...");
-                if (ForwardedPorts?.Any() == true)
-                {
-                    foreach (var pf in ForwardedPorts)
-                    {
-                        try
-                        {
-                            await pf.Disconnect();
-                        }
-                        catch (Exception ex)
-                        {
-                            MyLogger.Log("Failed to close a forwarded port.", ex);
-                        }
-                    }
-                    ForwardedPorts.ToList().ForEach(x => Children.Remove(x));
-                    this.WriteLog("Closed All Forwarded Ports.");
-                }
                 try
                 {
                     await Task.Run(() =>
                     {
-                        SshClient.Disconnect();
-                        SshClient.Dispose();
+                        SshClient?.Disconnect();
+                        SshClient?.Dispose();
                         SshClient = null;
                     });
                     this.WriteLog("Disconnected.");

